akka-cluster-sharding/src/main/resources/reference.conf

```diff
@@ -110,6 +110,16 @@ akka.cluster.sharding {
     max-simultaneous-rebalance = 3
   }
 
+  dynamic-shard-allocation-strategy {
+    # Dynamic allocation stored shardid -> address in a map in ddata. Shard across this many
+    # maps to avoid any one map getting too big when there is a large number of shards.
+    # Must be the same on every node.
+    ddata-keys = 5
```

Comment:
Yep, or do you think we should just hard code this and wait for a need for it to be configurable?

Reply:
I think that’s what we (I) did for remember entities. Less knobs to turn wrong. We should do some estimate/test of how many shards we can handle with X keys. (message size of ddata gossip)

Reply:
I actually see large `Gossip` messages when running with multiple keys.

Tested by adding many shard allocations and updating them to make the digests not match.

```
1 key, 100 shards
Gossip size 1283 compressed gossip size 1269
1 key, 1000 shards
Gossip size 9489 compressed gossip size 6717
1 key, 10000 shards
Gossip size 103859 compressed gossip size 52552
1 key, 100,000 shards
Gossip size 125490 compressed gossip size 62280

5 keys, 10,000 shards
Gossip size 124318 compressed gossip size 87293

5 keys, 100,000 shards
 Gossip size 408827 compressed gossip size 296319
```

When spreading 100,000 shards over 5 keys we hit the max message size.

Is this to do with the optimisations with have for primitive maps?

I'll take out the sharding across keys and just use a single one as supporting 100,000 shards/partitions seems plenty.

Reply:
spreading over keys might but not necessarily will reduce the gossip size: the replicator will first detect which keys have changes, and only gossip the keys for which there are changes. These gossip messages are batched: when there are multiple keys with changes, it is possible that multiple of those will be gossiped in a single message, though never more than akka.cluster.distributed-data.max-delta-elements elements at a time. The default value for this setting is 500.

It would be cool if we could set the max to a byte size rather than a number of elements size - but I guess to know that we'd have to actually serialize the objects so that's not so easy.

Reply:
> It would be cool if we could set the max to a byte size rather than a number of elements size

yes, that is proposed by @helena in #28421

Reply:
Aren't these results surprising?

> 1 key, 10000 shards
Gossip size 103859 compressed gossip size 52552
> 1 key, 100,000 shards
Gossip size 125490 compressed gossip size 62280

why doesn't it increase more when going from 10 to 100?

Reply:
I ran this a few times and was checking the serialized size of the Gossip message.  I'll sanity check the results again

Reply:
> spreading over keys might but not necessarily will reduce the gossip size: the replicator will first detect which keys have changes, and only gossip the keys for which there are changes. These gossip messages are batched: when there are multiple keys with changes, it is possible that multiple of those will be gossiped in a single message, though never more than akka.cluster.distributed-data.max-delta-elements elements at a time. The default value for this setting is 500.

Gotcha, thanks @raboof 

So the benefit would be that if we have a lot of keys it is unlikely that all of them will change at once so the gossip will be smaller. 

I think this is something we can change in the future as the allocation being in the wrong key is fine as long as the allocation strategy is reading them all. It just may have duplicates which we could deal with if we change later to use multiple keys.

Reply:
Sounds good to me!

Reply:
My test was stupid, i was making the updates but the messages were stacked up in the replicators mailbox and weren't all processed before the test ended.

I'm running a test now, currently at 33k entries with:

```
At 33000
[info] [JVM-2] Gossip size 314907 compressed size 176313 with Set(external-sharding-home)
```

This is increasing linearly with the number of keys. I think 10s of thousands of shards is quite unlikely but we can re-visit if

Reply:
ok, good to understand the numbers. I agree, starting simple is always better when it's possible to expand it later.